code/datums/craft/step.dm

```diff
@@ -107,8 +128,39 @@
 	if(istype(belt))
 		items += belt.contents
 
+	//Robots can use their module items as tools or materials
+	//We will do a check later to prevent them from dropping their tools as consumed components
+	if (isrobot(user))
+		var/mob/living/silicon/robot/R = user
+		if (R.module_state_1)
+			items += R.module_state_1
+		if (R.module_state_2)
+			items += R.module_state_2
+		if (R.module_state_3)
+			items += R.module_state_3
+
+	//We will allow all items in a 3x3 area, centred on the tile infront, to be used as components or mats
+	//Tools must be held though
+	if (!reqed_quality)
+		var/turf/T = get_step(user, user.dir)
+		//Use atom/movable to account for the possiblity of recipes requiring live or dead mobs as ingredients
+		for (var/atom/movable/A in range(1, T))
+			if (!A.anchored)
+				items += A
+
 	if(reqed_type)
-		return locate(reqed_type) in items
+		//Special handling for items that will be consumed
+		for(var/atom/movable/I in items)
+			//First we find the item
+			if (!istype(I, reqed_type))
+				//not the right type
+				continue
+			//Okay, so we found something that matches
+			if (is_valid_to_consume(I, user))
```

Comment:
That check not reqed if req_amount == 0 item will not be unequipped or consumed.

Reply:
yes it will. With ingredients the code always assumes one unit, and unequips and consumes it appropriately

Also there is at least one case (tray shield) where someone has tried writing in a quantity for an ingredient. Which doesn't do anything right now, but may cause issues if i tried a check like that

is_valid_to_consume is a generic safety check. Crafting code is non performance critical, so it's better to be robust and safe. we're likely to see a lot of submissions of crafting recipes by inexperienced coders in future